<a href="https://colab.research.google.com/github/arthurt1/Avalia-o-de-Modelos-de-ML/blob/rodrigo/Avalia%C3%A7%C3%A3o_de_Modelos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nomes da Equipe:



*   Arthur Thomé Costa - 565756
*   Rodrigo Sousa Barbosa - 566581
*   Ryan dos Santos Oliveira - 565120




# Objetivos:
*   Aplicar diferentes métricas de avaliação para classificação, regressão e clusterização
*  Comparar o desempenho de múltiplos algoritmos
* Interpretar corretamente os resultados das métricas
* Identificar problemas como overfitting e underfitting

# Pequeno resumo do trabalho e resultados:

Preparação do Ambiente Virtual:

In [6]:
## Importação das bibliotecas
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.datasets import load_iris, load_wine, fetch_openml
from sklearn.cluster import KMeans

# Parte 1: CLASSIFICAÇÃO


---


Dataset: *Iris Dataset*

Descrição: Classificação de espécies de flores Iris (Setosa, Versicolor, Virginica) baseado em medidas das sépalas e pétalas.

## Tarefa 1.1: Implementação e Avaliação Básica

Implemente o classificador K-Nearest Neighbors (KNN) com k=3 e k=5

### a) Divida o dataset em 70% treino e 30% teste. Para cada modelo, calcule:

- Acurácia
- Precisão (macro e micro)
- Recall (macro e micro)
- F1-score (macro e micro)
- Matriz de confusão

In [10]:
iris = load_iris()

df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [8]:
from sklearn.neighbors import KNeighborsClassifier


# Define features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Split the dataset into 70% training and 30% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("--- K-Nearest Neighbors (k=3) ---")

# Implement KNN with k=3
knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_3.fit(X_train_scaled, y_train)
y_pred_knn_3 = knn_3.predict(X_test_scaled)

# Calculate metrics for k=3
accuracy_knn_3 = accuracy_score(y_test, y_pred_knn_3)
precision_macro_knn_3 = precision_score(y_test, y_pred_knn_3, average='macro')
precision_micro_knn_3 = precision_score(y_test, y_pred_knn_3, average='micro')
recall_macro_knn_3 = recall_score(y_test, y_pred_knn_3, average='macro')
recall_micro_knn_3 = recall_score(y_test, y_pred_knn_3, average='micro')
f1_macro_knn_3 = f1_score(y_test, y_pred_knn_3, average='macro')
f1_micro_knn_3 = f1_score(y_test, y_pred_knn_3, average='micro')
confusion_matrix_knn_3 = confusion_matrix(y_test, y_pred_knn_3)

print(f"Accuracy (k=3): {accuracy_knn_3:.4f}")
print(f"Precision (Macro, k=3): {precision_macro_knn_3:.4f}")
print(f"Precision (Micro, k=3): {precision_micro_knn_3:.4f}")
print(f"Recall (Macro, k=3): {recall_macro_knn_3:.4f}")
print(f"Recall (Micro, k=3): {recall_micro_knn_3:.4f}")
print(f"F1-score (Macro, k=3): {f1_macro_knn_3:.4f}")
print(f"F1-score (Micro, k=3): {f1_micro_knn_3:.4f}")
print("Confusion Matrix (k=3):")
print(confusion_matrix_knn_3)

print("\n--- K-Nearest Neighbors (k=5) ---")

# Implement KNN with k=5
knn_5 = KNeighborsClassifier(n_neighbors=5)
knn_5.fit(X_train_scaled, y_train)
y_pred_knn_5 = knn_5.predict(X_test_scaled)

# Calculate metrics for k=5
accuracy_knn_5 = accuracy_score(y_test, y_pred_knn_5)
precision_macro_knn_5 = precision_score(y_test, y_pred_knn_5, average='macro')
precision_micro_knn_5 = precision_score(y_test, y_pred_knn_5, average='micro')
recall_macro_knn_5 = recall_score(y_test, y_pred_knn_5, average='macro')
recall_micro_knn_5 = recall_score(y_test, y_pred_knn_5, average='micro')
f1_macro_knn_5 = f1_score(y_test, y_pred_knn_5, average='macro')
f1_micro_knn_5 = f1_score(y_test, y_pred_knn_5, average='micro')
confusion_matrix_knn_5 = confusion_matrix(y_test, y_pred_knn_5)

print(f"Accuracy (k=5): {accuracy_knn_5:.4f}")
print(f"Precision (Macro, k=5): {precision_macro_knn_5:.4f}")
print(f"Precision (Micro, k=5): {precision_micro_knn_5:.4f}")
print(f"Recall (Macro, k=5): {recall_macro_knn_5:.4f}")
print(f"Recall (Micro, k=5): {recall_micro_knn_5:.4f}")
print(f"F1-score (Macro, k=5): {f1_macro_knn_5:.4f}")
print(f"F1-score (Micro, k=5): {f1_micro_knn_5:.4f}")
print("Confusion Matrix (k=5):")
print(confusion_matrix_knn_5)


--- K-Nearest Neighbors (k=3) ---
Accuracy (k=3): 1.0000
Precision (Macro, k=3): 1.0000
Precision (Micro, k=3): 1.0000
Recall (Macro, k=3): 1.0000
Recall (Micro, k=3): 1.0000
F1-score (Macro, k=3): 1.0000
F1-score (Micro, k=3): 1.0000
Confusion Matrix (k=3):
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]

--- K-Nearest Neighbors (k=5) ---
Accuracy (k=5): 1.0000
Precision (Macro, k=5): 1.0000
Precision (Micro, k=5): 1.0000
Recall (Macro, k=5): 1.0000
Recall (Micro, k=5): 1.0000
F1-score (Macro, k=5): 1.0000
F1-score (Micro, k=5): 1.0000
Confusion Matrix (k=5):
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]


### b) Qual modelo obteve melhor desempenho geral? Justifique sua resposta considerando múltiplas métricas.

Ambos os modelos K-Nearest Neighbors, com k=3 e k=5, obtiveram desempenho idêntico e perfeito em todas as métricas de avaliação (Acurácia, Precisão, Recall e F1-score, tanto macro quanto micro), além de apresentarem a mesma Matriz de Confusão. Isso significa que, para este conjunto de dados de teste específico, ambos os modelos foram capazes de classificar todas as instâncias corretamente, resultando em 100% de acerto. Portanto, não há um desempenho superior de um modelo sobre o outro neste cenário.

### c) Existe alguma classe que é mais difícil de classificar? Como você identificou isso?

Analisando as Matrizes de Confusão para ambos os modelos (k=3 e k=5):

```
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]
```

Podemos observar que todos os valores fora da diagonal principal são zero. Isso indica que não houve nenhuma classificação incorreta para nenhuma das classes (0, 1 ou 2). Todos os 19 exemplos da classe 0, os 13 exemplos da classe 1 e os 13 exemplos da classe 2 foram corretamente identificados.

Portanto, para este conjunto de dados de teste, **não existe nenhuma classe que tenha sido mais difícil de classificar**, pois ambos os modelos alcançaram uma precisão perfeita para todas as espécies de Iris.

## Tarefa 1.2: Validação Cruzada

### a)Implemente validação cruzada k-fold com k=5 para todos os modelos da Tarefa 1.1.

### b)Compare os resultados da validação cruzada com a divisão simples treino/teste. O que você observa?

### c) Calcule o desvio padrão da acurácia para cada modelo. O que isso indica sobre a estabilidade do modelo?

## Tarefa 1.3: Curvas ROC e AUC

### a) Para o problema binário (Setosa vs. Não-Setosa), plote as curvas ROC para os três modelos.

### b) Calcule a área sob a curva (AUC) para cada modelo.

### c) Interprete os resultados: qual modelo tem melhor capacidade de discriminação?

# Parte 2: REGRESSÃO

---

Dataset: *Boston Housing (ou California Housing)*

Descrição: Predição de preços de imóveis baseado em características socioeconômicas e geográficas.

## Tarefa 2.1: Múltiplos Modelos de Regressão

Implemente o modelo. K-Nearest Neighbors Regressor (k=3 e k=7)

### a) Divida o dataset em 70% treino e 30% teste. Para cada modelo, calcule:

- Erro Médio Absoluto (MAE)
- Erro Quadrático Médio (MSE)
- Raiz do Erro Quadrático Médio (RMSE)
- Coeficiente de Determinação (R²)
- Erro Percentual Absoluto Médio (MAPE)

### b) Crie um gráfico de dispersão comparando valores preditos vs. valores reais para cada modelo.

### c) Qual modelo apresentou melhor desempenho? Por quê?

## Tarefa 2.2: Análise de Resíduos

### a) Para os três melhores modelos, plote o gráfico de resíduos (erro vs. valores preditos).

### b) Há algum padrão nos resíduos? O que isso indica sobre o modelo?

### c) Calcule e plote o histograma dos resíduos. Eles seguem uma distribuição aproximadamente normal?

# Parte 3: CLUSTERIZAÇÃO


---


Dataset: *Wine Dataset*

Descrição: Análise química de vinhos de três cultivares diferentes na Itália.

## Tarefa 3.1: K-Means

### a) Aplique o algoritmo K-Means com k variando de 2 a 10.

In [ ]:
wine = load_wine()
X = wine.data

## Padronização
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


## K-Means variando k de 2 a 10
resultados = []

for k in range(2, 11):
    kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
    kmeans.fit(X_scaled)

    inertia = kmeans.inertia_  # SSE (erro interno)
    sil = silhouette_score(X_scaled, kmeans.labels_)

    resultados.append([k, inertia, sil])

## mostrar resultados em tabela
df_result = pd.DataFrame(resultados, columns=["k", "Inertia (SSE)", "Silhouette"])
print(df_result)

    k  Inertia (SSE)  Silhouette
0   2    1658.758852    0.259317
1   3    1277.928489    0.284859
2   4    1175.428333    0.260170
3   5    1109.512739    0.201619
4   6    1046.002333    0.237167
5   7     981.595233    0.203628
6   8     935.201211    0.157014
7   9     889.892911    0.149882
8  10     845.895237    0.143638


### b) Para cada valor de k, calcule:

- Inércia (within-cluster sum of squares)
- Silhouette Score
- Davies-Bouldin Index
- Calinski-Harabasz Index

### c) Plote o Método do Cotovelo (Elbow Method) usando a inércia.

### d) Determine o número ótimo de clusters considerando múltiplas métricas.

## Tarefa 3.2: Comparação de Algoritmos
Implemente e compare:

1. K-Means (com k ótimo da Tarefa 3.1)

2. DBSCAN (teste diferentes valores de eps e min_samples)

3. Hierarchical Clustering (Agglomerative)

### a) Calcule o Silhouette Score para cada algoritmo.

### b) Como os labels verdadeiros (cultivares) se comparam com os clusters encontrados?

## Tarefa 3.3: Análise de Perfis dos Clusters

### a) Para a melhor solução de clusterização, calcule:

- Média de cada feature por cluster
- Desvio padrão de cada feature por cluster

### b) Crie um gráfico de radar (spider chart) mostrando o perfil médio de cada cluster.

### c) Descreva as características de cada cluster em termos das features originais.